### DISCUSSION TITLE
Plugin for Zebra Label Printer

### DISCUSSION THREAD

**Discussion by SergeoLacruz** [inventree/InvenTree#3202] [Q&A] [answered]
Hi,
I created a simple plugin for a Zebra GK420T and wrote some doc for it:

https://github.com/SergeoLacruz/inventree-zebra-plugin

Two questions from my side remain:
I case the printer cannot be reached a raise an exception and get the following output in the worker log:

 ```code
Error with the connection
Label printing failed: Sending notification to user 'admin'
/home/michael/.local/lib/python3.9/site-packages/django/db/models/fields/__init__.py:1358: RuntimeWarning: DateTimeField NotificationEntry.updated received a naive datetime (2022-06-14 00:00:00) while time zone support is active.
 warnings.warn("DateTimeField %s.%s received a naive datetime "
```
The printer error shows up in the admin interface but only sometimes in the users notification inbox.

I created a label html with 50mmx30mm. From the plugin API i get a png of 615x378 which relates to 312dpi. Where is this defined and can it be changed? Thanks in advance,

Michael

**Comment by matmair** [comment]
Hi there @SergeoLacruz !
Thank you for your effort. We will make sure to promote this once your questions have been solved. Regarding the first one: That seems to be a dateline issue - I will look into it later today (UTC +2)
The second one seems to be about printers - @schrodingersgat is the expert there.

**Comment by SchrodingersGat** [comment]
> The printer error shows up in the admin interface but only sometimes in the users notification inbox.

There is a rate limit on user notifications to prevent flooding the notification system. This is probably what you are encountering here.

**Reply by SergeoLacruz** [reply]
In this case the rate limit is about one message a day or even less. I assume I have a different problem. Can I remove the rate limit for debugging?

**Reply by SchrodingersGat** [reply]
Not easily. I'd suggest you just navigate to the "errors" page in the admin interface, and you can see all the error messages there.

https://inventree.readthedocs.io/en/latest/settings/logs/

**Reply by SergeoLacruz** [reply]
This I can do. 
I am just thinking about  a normal user who has no admin access. He just sees that the printer does not print but does not receive any message. I play with my system regularly and saw the last notification three days ago.

**Reply by matmair** [reply]
@SergeoLacruz you could also just raise a notification - this will be sent via any supported method.

**Comment by SchrodingersGat** [comment, marked as answer]
> I created a label html with 50mmx30mm. From the plugin API i get a png of 615x378 which relates to 312dpi. Where is this defined and can it be changed? 

Currently the DPI is hard-coded, but I'm not entirely happy with the approach I used here.

https://github.com/inventree/InvenTree/blob/e83995b4f564574445e2ba059de0c03ed69411c4/InvenTree/plugin/base/label/label.py#L41

There are some options here:

- We can expose the DPI as a configurable setting (Edit: https://github.com/inventree/InvenTree/pull/3208)
- You can manually generate an image at whatever DPI you want, as the original PDF data are also provided.

**Reply by SergeoLacruz** [reply]
Having it hard coded is OK for me even if 312 in not 300. I have another scaling possibility when I convert from pillow to zebra. I just wanted to make sure that it does not change under certain conditions and garbles my output.

**Comment by SergeoLacruz** [comment]
Hi,
I fixed some scaling problems with the label sizes and added a few words to the doc how the scaling works. There are certain things that could be done better but for me it is OK as it is. Comments invited. 
Michael

**Reply by matmair** [reply]
I do not have a Zebra printer so i can not test it sadly.
You have a very clean way to code / work - really liking it! Would it be ok for you if we put a link to your plugin in the docs? Possibly here: https://inventree.readthedocs.io/en/latest/extend/integrate/ but maybe I will create a separate plugin section.

**Reply by SergeoLacruz** [reply]
Hi,
Thank you. There was not much code to write. Basically I copied the brother plugin and changed it accordingly.  It was tested with a real GK420t. I added a picture of the test setup :-). You are free to to set a link in the docs. 
Michael

**Reply by matmair** [reply]
PR is in: https://github.com/inventree/inventree-docs/pull/303

**Reply by SergeoLacruz** [reply]
I did some updates to the zebra plugin to fit the current api changes. Anyhow it does not work any more with   | 0.9.0 dev.
Everything seems fine but the worker output tells me: Could not find matching plugin for 'zebra'

Michael

**Reply by SchrodingersGat** [reply]
@SergeoLacruz please open a new issue to discuss this

**Comment by SergeoLacruz** [comment]
I did some update on the plugin:

https://github.com/SergeoLacruz/inventree-zebra-plugin

Added parameters for label width and height. I also found the parameter validators and added them:-) Now adding a non numerical value shows a form error. 
The MinValueValidator and MaxValueValidator do not seem to work. At least adding -1 for width is still accepted. Any idea?

Michael

**Reply by SchrodingersGat** [reply]
@SergeoLacruz not sure, the validators *should* be working. Does the maximum limit work?

**Comment by SergeoLacruz** [comment]
No,
Tried with threshold (Validators set to 0 and 255):
Typing a string results in form error
Typing a decimal result in fallback to default
-100 is accepted as well as 1500
![validator](https://user-images.githubusercontent.com/10836099/227716591-c9949503-878d-46a5-8603-bbac19d958fe.png)

**Reply by SchrodingersGat** [reply]
That's odd, these validators definitely work for "non plugin" settings. You may have to do some local debugging :)

**Comment by SergeoLacruz** [comment]
I did some update on the driver. Added features that where needed for different label sizes and printers. A also added a section on scaling, darkness and quality of QR-codes to the docs. Comments invited.

**Comment by SergeoLacruz** [comment]
I did some update on the plugin:

https://github.com/SergeoLacruz/inventree-zebra-plugin

I turned out that the ZPL librara does not send important commands for the label size. So the labels where odd if another print job changeed these parameters. I added some lines to the zpl lib to fix this. 

Zebra printers can remember these settings. If a job rotates the label, all following labels come out rotated unless there are commands to change this in the current job. For this reason I added a parameter that allows to add raw ZPL commands to the printout. These an be used to bring the printer to the correct state. 

I am still struggling with errors. After pushing print InvenTree shows a green bar telling "Label sent to printer" even if the print fails. I found no way yet to change this to a red bar telling "Printing failed" Is this possible?

Michael
